### PR TITLE
Wire through SIGTERM/KILL bound stopCh to API servers for graceful shutdown

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

**Comment by jim-minter** [comment]
Apologies @sttts I don't have the experience to be able to lgtm this.  @deads2k ?

File: pkg/cmd/openshift-apiserver/cmd.go [hunk 1]
```diff
@@ -90,19 +89,13 @@ func (o *OpenShiftAPIServer) Validate() error {
 	return nil
 }
 
-// StartAPIServer calls RunAPIServer and then waits forever
-func (o *OpenShiftAPIServer) StartAPIServer() error {
+func (o *OpenShiftAPIServer) WithoutNetworkingAPI() *OpenShiftAPIServer {
```
Comment (soltysh) [inline pkg/cmd/openshift-apiserver/cmd.go]: This is good, I prefer explicit stuff!

File: test/util/server/testing.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	configapi "github.com/openshift/origin/pkg/cmd/server/apis/config"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func StartConfiguredMaster(masterConfig *configapi.MasterConfig) (string, error) {
+	if v := flag.Lookup("test.v"); v == nil {
+		panic("cannot be used outside of test code")
+	}
+
+	// this is for testing. TODO: update the billion call sites with proper stopCh setup code
```
Comment (soltysh) [inline test/util/server/testing.go]: That's nasty :stuck_out_tongue_closed_eyes:  but I'll be cleaning that soon-ish.

File: pkg/cmd/server/start/start_master.go [hunk 3]
```diff
@@ -218,21 +218,23 @@ func (o MasterOptions) StartMaster() error {
 	// TODO: this should be encapsulated by RunMaster, but StartAllInOne has no
 	// way to communicate whether RunMaster should block.
 	go daemon.SdNotify(false, "READY=1")
```
Comment (mfojtik) [inline pkg/cmd/server/start/start_master.go]: @sttts probably time to nuke this ;-)

Comment (sttts) [reply pkg/cmd/server/start/start_master.go]: it's openshift start anyway. That will die soon.

Comment (sttts) [reply pkg/cmd/server/start/start_master.go]: And btw, genericapiserver calls the same.

**Comment by smarterclayton** [comment]
I wish I’d reviewed earlier, but passing context down would probably be better in the long run (we want to pass contexts deeper).  Also note for us to get to graceful shutdown from an http server we need to do these things:

1. Wait X seconds
2. Call http.Server.Shutdown
3. As soon as all connections are down we can exit OR hard timeout

**Comment by soltysh** [comment]
> I wish I’d reviewed earlier, but passing context down would probably be better in the long run (we want to pass contexts deeper)

I was considering that, given that rebase is switching to it. It might be that when I pick this change in the rebase I'll have to change to to context, though.

**Comment by soltysh** [comment]
But let's land this asap, and we can iterate further later.

**Comment by smarterclayton** [comment]
Context can wait until we have context in client go.  Once we have contexts in more places we’ll want to close the loop because context is usually superior and we can carry trace info down.

**Comment by sttts** [comment]
Move last commit to https://github.com/openshift/origin/pull/21594 and reapplying old label.

**Pull request merged**